src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash3.cs

```diff
@@ -0,0 +1,932 @@
+
+                ref uint sourceRef = ref Unsafe.As<byte, uint>(ref MemoryMarshal.GetReference(source));
+                ref uint keyRef = ref Unsafe.As<byte, uint>(ref MemoryMarshal.GetReference(secret));
+
+                while (Unsafe.IsAddressLessThan(ref accRef, ref accEnd))
+                {
+                    Vector256<ulong> accVec = Vector256.LoadUnsafe(ref accRef);
+                    Vector256<uint> sourceVec = Vector256.LoadUnsafe(ref sourceRef);
+                    Vector256<uint> sourceKey = sourceVec ^ Vector256.LoadUnsafe(ref keyRef);
+
+                    // TODO: Figure out how to unwind this shuffle and just use Vector256.Multiply
+                    Vector256<uint> sourceKeyLow = Vector256.Shuffle(sourceKey, Vector256.Create(1u, 0, 3, 0, 5, 0, 7, 0));
+                    Vector256<uint> sourceSwap = Vector256.Shuffle(sourceVec, Vector256.Create(2u, 3, 0, 1, 6, 7, 4, 5));
+                    Vector256<ulong> sum = accVec + sourceSwap.AsUInt64();
+                    Vector256<ulong> product = Avx2.IsSupported ?
+                        Avx2.Multiply(sourceKey, sourceKeyLow) :
+                        (sourceKey & Vector256.Create(~0u, 0u, ~0u, 0u, ~0u, 0u, ~0u, 0u)).AsUInt64() * (sourceKeyLow & Vector256.Create(~0u, 0u, ~0u, 0u, ~0u, 0u, ~0u, 0u)).AsUInt64();
+
+                    Vector256.StoreUnsafe(product + sum, ref accRef);
+
+                    accRef = ref Unsafe.Add(ref accRef, Vector256<ulong>.Count);
+                    keyRef = ref Unsafe.Add(ref keyRef, Vector256<uint>.Count);
+                    sourceRef = ref Unsafe.Add(ref sourceRef, Vector256<uint>.Count);
+                }
+
+                Debug.Assert(!Unsafe.IsAddressGreaterThan(ref accRef, ref Unsafe.Add(ref accRef, acc.Length)));
+                Debug.Assert(!Unsafe.IsAddressGreaterThan(ref sourceRef, ref Unsafe.Add(ref sourceRef, source.Length / sizeof(uint))));
+                Debug.Assert(!Unsafe.IsAddressGreaterThan(ref keyRef, ref Unsafe.Add(ref keyRef, secret.Length / sizeof(uint))));
+            }
+            else if (Vector128.IsHardwareAccelerated && BitConverter.IsLittleEndian)
+            {
+                ref ulong accRef = ref MemoryMarshal.GetReference(acc);
+                ref ulong accEnd = ref Unsafe.Add(ref accRef, acc.Length);
+
+                ref uint sourceRef = ref Unsafe.As<byte, uint>(ref MemoryMarshal.GetReference(source));
+                ref uint keyRef = ref Unsafe.As<byte, uint>(ref MemoryMarshal.GetReference(secret));
+
+                while (Unsafe.IsAddressLessThan(ref accRef, ref accEnd))
+                {
+                    Vector128<ulong> accVec = Vector128.LoadUnsafe(ref accRef);
+                    Vector128<uint> sourceVec = Vector128.LoadUnsafe(ref sourceRef);
+                    Vector128<uint> sourceKey = sourceVec ^ Vector128.LoadUnsafe(ref keyRef);
+
+                    // TODO: Figure out how to unwind this shuffle and just use Vector128.Multiply
```

Comment:
@tannergooding, per our conversation yesterday, I was able to get rid of the other direct usage of Sse2.Multiply with a shuffle done before it, but I couldn't quite wrap my mind around getting rid of this one.  If it's obvious to you how, please let me know :)